pkg/asset/installconfig/gcp/usertags.go

```diff
@@ -0,0 +1,298 @@
+	projectParentPathFmt = "//cloudresourcemanager.googleapis.com/projects/%s"
+)
+
+// effectiveTags is the custom type for holding tags.
+type effectiveTags map[string]struct{}
+
+// processedUserTags is for storing the validated and processed tags
+// in-memory for later use. Validating and converting the tags to
+// required format involves using same GCP APIs and to reduce the
+// number of calls, making use of the in-memory storage approach.
+type processedUserTags struct {
+	processed bool
+	tags      map[string]string
+
+	sync.Mutex
+}
+
+// processedTags is the global instance for storing the processed tags.
+var processedTags = newProcessedUserTags()
+
+// tagManager handles resource tagging.
+type tagManager struct {
+	client API
+}
+
+// TagManager is the interface that wraps methods for resource tag operations.
+type TagManager interface {
+	getTagClientOptions() []option.ClientOption
+}
+
+// tagValuesClient handles operations on tag value resources.
+type tagValuesClient struct {
+	*rscmgr.TagValuesClient
+}
+
+// tagValuesClient handles operations on tag binding resources.
+type tagBindingsClient struct {
+	*rscmgr.TagBindingsClient
+}
+
+// newProcessedUserTags is for initializing an instance of processedUserTags.
+func newProcessedUserTags() *processedUserTags {
+	return &processedUserTags{}
+}
+
+// make is for initializing the internals of processedUserTags.
+func (p *processedUserTags) make(size int) {
+	p.Lock()
+	defer p.Unlock()
+
+	p.tags = make(map[string]string, size)
+}
+
+// addTag is for adding a tag-set to processedUserTags.
+func (p *processedUserTags) addTag(k, v string) {
+	p.Lock()
+	defer p.Unlock()
+
+	p.tags[k] = v
+}
+
+// updateState is for updating the tags processed status.
+func (p *processedUserTags) updateState() {
+	p.Lock()
+	defer p.Unlock()
+
+	p.processed = true
+}
+
+// getState is for getting the tags processed status.
+func (p *processedUserTags) getState() bool {
+	p.Lock()
+	defer p.Unlock()
+
+	return p.processed
+}
+
+// copy is for making copy of the tags in processedUserTags.
+func (p *processedUserTags) copy() map[string]string {
+	p.Lock()
+	defer p.Unlock()
+
+	t := make(map[string]string, len(p.tags))
+	for k, v := range p.tags {
+		t[k] = v
+	}
+	return t
+}
+
+// NewTagManager creates a tagManager instance.
+func NewTagManager(client API) TagManager {
+	return &tagManager{client: client}
+}
+
+// GetUserTags returns the processed list of user provided tags if already available,
+// else validates, stores and returns the processed tags.
+func GetUserTags(mgr TagManager, ic *types.InstallConfig) (map[string]string, error) {
+	return getProcessedTags(mgr, ic)
+}
+
+// getProcessedTags returns the list user-provided validated tags.
+func getProcessedTags(mgr TagManager, ic *types.InstallConfig) (map[string]string, error) {
+	if !processedTags.getState() {
+		if err := validateUserTags(mgr, ic); err != nil {
```

Comment:
It's a bit weird that the definition for `validateUserTags` lives in the `validation.go` file and it basically just calls `validateAndStoreUserTags` which is defined in this file. Also, I don't see why this function has to receive the `InstallConfig` object if it only needs the project name and the userTags to do its job.

Reply:
Yeah, I missed this unnecessary call while incorporating the comments, I will update it.
I thought to pass the `InstallConfig` object reference and use what's required, instead of passing individual parameters. Please let me know, if needs to be changed.